src/main/java/gregtech/api/capability/impl/CleanroomLogic.java

```diff
@@ -0,0 +1,228 @@
+    }
+
+    /**
+     * Performs the actual cleaning
+     * Call this method every tick in update
+     */
+    public void updateLogic() {
+        if (metaTileEntity.getWorld().isRemote) return;
+
+        if (hasMaintenance && ((IMaintenance) metaTileEntity).getNumMaintenanceProblems() > 5) {
+            return;
+        }
+
+        // cleanrooms which cannot work do nothing
+        if (!this.isWorkingEnabled) return;
+
+        // drain the energy
+        consumeEnergy(false);
+
+        if (!this.isActive) setActive(true);
+
+        // increase progress
+        progressTime++;
+        if (progressTime % getMaxProgress() != 0) return;
+        progressTime = 0;
+
+        int amountToClean;
+        if (hasEnoughPower()) amountToClean = BASE_CLEAN_AMOUNT * (getTierDifference() + 1);
+        else amountToClean = -BASE_CLEAN_AMOUNT;
+
+        // each maintenance problem lowers gain by 1 and increases loss by 1
+        if (hasMaintenance) amountToClean -= ((IMaintenance) metaTileEntity).getNumMaintenanceProblems();
+        ((ICleanroomProvider) metaTileEntity).adjustCleanAmount(amountToClean);
+    }
+
+    protected boolean consumeEnergy(boolean simulate) {
+        return ((ICleanroomProvider) metaTileEntity).drainEnergy(simulate);
+    }
+
+    /**
+     * @return true if the cleanroom is able to drain energy, else false
+     */
+    protected boolean hasEnoughPower() {
+        if (!consumeEnergy(true)) {
+            if (progressTime >= 2) {
+                if (ConfigHolder.machines.recipeProgressLowEnergy) this.progressTime = 1;
+                else this.progressTime = Math.max(1, progressTime - 2);
+
+                hasNotEnoughEnergy = true;
+            }
+            return false;
+        }
+
+        if (this.hasNotEnoughEnergy && ((ICleanroomProvider) metaTileEntity).getEnergyInputPerSecond() > 19L * GTValues.VA[((ICleanroomProvider) metaTileEntity).getEnergyTier()]) {
+            this.hasNotEnoughEnergy = false;
+        }
+
+        return true;
+    }
+
+    /**
+     * @return true if the cleanroom is active
+     */
+    public boolean isActive() {
+        return this.isActive;
+    }
+
+    /**
+     * @param active the new state of the cleanroom's activity: true to change to active, else false
+     */
+    public void setActive(boolean active) {
+        if (this.isActive != active) {
+            this.isActive = active;
+            this.metaTileEntity.markDirty();
+            if (metaTileEntity.getWorld() != null && !metaTileEntity.getWorld().isRemote) {
+                this.metaTileEntity.writeCustomData(GregtechDataCodes.IS_WORKING, buf -> buf.writeBoolean(active));
+            }
+        }
+    }
+
+    /**
+     * @param workingEnabled the new state of the cleanroom's ability to work: true to change to enabled, else false
+     */
+    public void setWorkingEnabled(boolean workingEnabled) {
+        this.isWorkingEnabled = workingEnabled;
+        metaTileEntity.markDirty();
```

Comment:
This also needs to be synced via writeCustomData, otherwise we will have the issue that we had with the miners not properly syncing stuff